File: app/schemas/generation_batch.py
from pydantic import BaseModel, Field, model_validator
from typing import Any, Literal
from datetime import datetime

from app.schemas.test_case._core import TestCaseStep


class GenerationBatchCreate(BaseModel):
    project_id: int = Field(..., gt=0)
    entry_type: Literal["NEW_FEATURE_GENERATION", "HISTORY_UPDATE"] = "NEW_FEATURE_GENERATION"
    scenario_type: Literal[
        "B1_REQUIREMENT_TESTPOINT",
        "A1_REQUIREMENT_TESTPOINT_UI",
        "A2_HISTORY_EXCEL_REQUIREMENT_UI",
        "A3_HISTORY_XMIND_REQUIREMENT_UI",
        "B2_HISTORY_XMIND_UI",
        "B3_HISTORY_EXCEL_UI",
    ]
    generation_strategy: Literal[
        "REQUIREMENT_TESTPOINT_STANDARD_GENERATION",
        "FULL_CONTEXT_GENERATION_LITE",
        "HISTORY_INCREMENTAL_UPDATE",
        "HISTORY_UI_ADAPTATION",
    ]
    requirement_file_ids: list[int] = Field(default_factory=list, max_length=200)
    test_point_ids: list[int] = Field(default_factory=list, max_length=200)
    ui_screen_ids: list[int] = Field(default_factory=list, max_length=200)
    history_asset_ids: list[int] = Field(default_factory=list, max_length=200)
    client_request_id: str | None = Field(None, max_length=80)


class GenerationBatchUpdate(BaseModel):
    status: Literal[
        "created",
        "context_ready",
        "generating",
        "preview_ready",
        "saving",
        "saved",
        "partial_saved",
        "failed",
    ] | None = None
    context_stats: dict[str, Any] | None = None
    warnings: list[dict[str, Any]] | None = None
    evidence_refs: dict[str, Any] | None = None
    quality_summary: dict[str, Any] | None = None


class GenerationBatchResponse(BaseModel):
    id: int
    batch_no: str
    project_id: int
    user_id: int
    entry_type: str
    scenario_type: str
    generation_strategy: str
    status: str
    requirement_file_ids: list[int]
    test_point_ids: list[int]
    ui_screen_ids: list[int]
    history_asset_ids: list[int]
    context_stats: dict[str, Any]
    warnings: list[dict[str, Any]]
    evidence_refs: dict[str, Any]
    quality_summary: dict[str, Any]
    created_at: datetime
    updated_at: datetime | None

    model_config = {"from_attributes": True}


class PreviewCasePayload(BaseModel):
    client_id: str = Field(..., min_length=1, max_length=80)
    source_test_point_id: int | None = None
    requirement_file_id: int | None = None
    title: str = Field("", max_length=255)
    module: str = Field("", max_length=100)
    precondition: str = ""
    steps: list[TestCaseStep] = Field(default_factory=list)
    expected_result: str = ""
    priority: int = Field(2, ge=1, le=3)
    case_type: str = Field("manual", max_length=20)
    case_category: str | None = None
    quality_status: Literal["passed", "warning", "pending_review", "rejected"] = "pending_review"
    quality_issues: list[dict[str, Any]] = Field(default_factory=list)
    selected_for_save: bool = True
    source_refs: dict[str, Any] = Field(default_factory=dict)
    classification: str | None = Field(None, max_length=30, comment="历史更新分类: REUSE_CASE/UPDATE_CASE/NEW_CASE/DEPRECATED_CASE/CONFIRM_REQUIRED")
    history_case_id: int | None = Field(None, comment="UPDATE_CASE 对应的系统用例ID")
    update_action: Literal["create_new", "update_existing", "skip", "deprecate"] | None = Field(None, comment="保存动作")
    diff_fields: dict[str, Any] | None = Field(None, comment="变更字段及前后值")

    @model_validator(mode="after")
    def _validate_non_deprecate_fields(self) -> "PreviewCasePayload":
        if self.update_action == "deprecate":
            return self
        if not self.title:
            raise ValueError("非废弃场景 title 不能为空")
        if not self.steps:
            raise ValueError("非废弃场景 steps 不能为空")
        if not self.expected_result:
            raise ValueError("非废弃场景 expected_result 不能为空")
        return self


class GenerationBatchSaveRequest(BaseModel):
    idempotency_key: str = Field(..., min_length=1, max_length=100)
    save_mode: Literal["draft", "formal", "passed_only"]
    cases: list[PreviewCasePayload] = Field(..., min_length=1, max_length=500)


class BatchSaveFailureItem(BaseModel):
    client_id: str | None = None
    title: str | None = None
    reason: str


class GenerationBatchSaveResponse(BaseModel):
    batch_id: int
    idempotency_key: str
    save_mode: str
    saved_count: int
    failed_count: int
    saved_case_ids: list[int]
    failures: list[BatchSaveFailureItem]
    status: Literal["saved", "partial_saved", "failed"]

File: src/api/generationBatch.ts
import request from '@/utils/request'

export interface GenerationBatchCreatePayload {
  project_id: number
  entry_type: 'NEW_FEATURE_GENERATION'
  scenario_type: 'B1_REQUIREMENT_TESTPOINT' | 'A1_REQUIREMENT_TESTPOINT_UI'
  generation_strategy: 'REQUIREMENT_TESTPOINT_STANDARD_GENERATION' | 'FULL_CONTEXT_GENERATION_LITE'
  requirement_file_ids: number[]
  test_point_ids: number[]
  ui_screen_ids: number[]
  client_request_id?: string
}

export interface GenerationBatchResponse {
  id: number
  batch_no: string
  project_id: number
  user_id: number
  entry_type: string
  scenario_type: string
  generation_strategy: string
  status: string
  requirement_file_ids: number[]
  test_point_ids: number[]
  ui_screen_ids: number[]
  context_stats: Record<string, unknown>
  warnings: NormalizedWarning[]
  evidence_refs: Record<string, unknown>
  quality_summary: Record<string, unknown>
  created_at: string
  updated_at: string | null
}

export interface NormalizedWarning {
  code: string
  message: string
  detail: Record<string, unknown>
}

export interface GenerationBatchUpdatePayload {
  status?: string
  context_stats?: Record<string, unknown>
  warnings?: NormalizedWarning[]
  evidence_refs?: Record<string, unknown>
  quality_summary?: Record<string, unknown>
}

export interface PreviewCasePayload {
  client_id: string
  source_test_point_id?: number | null
  requirement_file_id?: number | null
  title: string
  module?: string
  precondition?: string
  steps: PreviewStepPayload[]
  expected_result: string
  priority?: number
  case_type?: string
  case_category?: string | null
  quality_status?: 'passed' | 'warning' | 'pending_review' | 'rejected'
  quality_issues?: Record<string, unknown>[]
  selected_for_save?: boolean
  source_refs?: Record<string, unknown>
  classification?: string | null
  history_case_id?: number | null
  update_action?: 'create_new' | 'update_existing' | 'skip' | 'deprecate' | null
  diff_fields?: Record<string, { old: unknown; new: unknown }> | null
}

export interface PreviewStepPayload {
  step: string | number
  action: string
  param?: string
  expected_result?: string | null
  test_data?: Record<string, unknown> | Record<string, unknown>[] | null
  description?: string | null
  ui_elements?: (string | Record<string, unknown>)[] | null
  action_type?: string | null
  input_value?: string | null
  target_element?: string | null
}

export interface GenerationBatchSavePayload {
  idempotency_key: string
  save_mode: 'draft' | 'formal' | 'passed_only'
  cases: PreviewCasePayload[]
}

export interface BatchSaveFailureItem {
  client_id?: string | null
  title?: string | null
  reason: string
}

export interface GenerationBatchSaveResponse {
  batch_id: number
  idempotency_key: string
  save_mode: string
  saved_count: number
  failed_count: number
  saved_case_ids: number[]
  failures: BatchSaveFailureItem[]
  status: 'saved' | 'partial_saved' | 'failed'
}

function extractData<T>(response: unknown): T {
  const resp = response as { code?: number; data?: T }
  if (resp && resp.data !== undefined) return resp.data
  return response as T
}

export const generationBatchApi = {
  create: async (data: GenerationBatchCreatePayload): Promise<GenerationBatchResponse> => {
    const response = await request.post('/api/v1/generation-batches', data)
    return extractData<GenerationBatchResponse>(response)
  },

  get: async (batchId: number): Promise<GenerationBatchResponse> => {
    const response = await request.get(`/api/v1/generation-batches/${batchId}`)
    return extractData<GenerationBatchResponse>(response)
  },

  update: async (batchId: number, data: GenerationBatchUpdatePayload): Promise<GenerationBatchResponse> => {
    const response = await request.patch(`/api/v1/generation-batches/${batchId}`, data)
    return extractData<GenerationBatchResponse>(response)
  },

  save: async (batchId: number, data: GenerationBatchSavePayload): Promise<GenerationBatchSaveResponse> => {
    const response = await request.post(`/api/v1/generation-batches/${batchId}/save`, data)
    return extractData<GenerationBatchSaveResponse>(response)
  },
}
